File: CD/3Add.py
class ThreeAddressCodeGenerator:
    def __init__(self):
        self.quadruples = []
        self.triples = []
        self.indirect_triples = []
        self.temp_count = 0

    def generate_temp(self):
        temp_var = f't{self.temp_count}'
        self.temp_count += 1
        return temp_var

    def generate_code(self, op, arg1, arg2, result):
        self.quadruples.append((op, arg1, arg2, result))
        self.triples.append((op, arg1, arg2))
        self.indirect_triples.append((op, result))

    def generate_expression_code(self, expr):
        if isinstance(expr, tuple):  # Binary operation
            op, left, right = expr
            temp_result = self.generate_temp()
            self.generate_expression_code(left)
            self.generate_expression_code(right)
            self.generate_code(op, left, right, temp_result)
            return temp_result
        else:  # Operand (variable or constant)
            return expr

    def display_code(self):
        print("Quadruples:")
        for quad in self.quadruples:
            print(quad)

        print("\nTriples:")
        for triple in self.triples:
            print(triple)

        print("\nIndirect Triples:")
        for indirect_triple in self.indirect_triples:
            print(indirect_triple)


def main():
    generator = ThreeAddressCodeGenerator()

    # Example expression: a + (b * c)
    expression = ('+', 'a', ('*', 'b', 'c'))

    result = generator.generate_expression_code(expression)

    generator.display_code()


if __name__ == "__main__":
    main()

File: CD/3Add.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <tuple>

using namespace std;

class ThreeAddressCodeGenerator {
public:
    vector<tuple<string, string, string, string>> quadruples;
    vector<tuple<string, string, string>> triples;
    vector<tuple<string, string>> indirect_triples;
    int temp_count;

    ThreeAddressCodeGenerator() : temp_count(0) {}

    string generate_temp() {
        return "t" + to_string(temp_count++);
    }

    void generate_code(const string& op, const string& arg1, const string& arg2, const string& result) {
        quadruples.push_back(make_tuple(op, arg1, arg2, result));
        triples.push_back(make_tuple(op, arg1, arg2));
        indirect_triples.push_back(make_tuple(op, result));
    }

    string generate_expression_code(const tuple<string, string, string>& expr) {
        string op = get<0>(expr);
        string left = get<1>(expr);
        string right = get<2>(expr);

        if (op == "+" || op == "-" || op == "*" || op == "/") {
            string temp_result = generate_temp();
            generate_expression_code(make_tuple(left, "", ""));
            generate_expression_code(make_tuple(right, "", ""));
            generate_code(op, left, right, temp_result);
            return temp_result;
        } else {
            // Operand (variable or constant)
            return op;
        }
    }

    void display_code() {
        cout << "Quadruples:\n";
        for (const auto& quad : quadruples) {
            cout << "(" << get<0>(quad) << ", " << get<1>(quad) << ", " << get<2>(quad) << ", " << get<3>(quad) << ")\n";
        }

        cout << "\nTriples:\n";
        for (const auto& triple : triples) {
            cout << "(" << get<0>(triple) << ", " << get<1>(triple) << ", " << get<2>(triple) << ")\n";
        }

        cout << "\nIndirect Triples:\n";
        for (const auto& indirect_triple : indirect_triples) {
            cout << "(" << get<0>(indirect_triple) << ", " << get<1>(indirect_triple) << ")\n";
        }
    }
};

int main() {
    ThreeAddressCodeGenerator generator;

    // Example expression: a + (b * c)
    tuple<string, string, string> expression = make_tuple("+", "a", "*");
    get<2>(expression) = generator.generate_expression_code(make_tuple("b", "*", "c"));

    string result = generator.generate_expression_code(expression);

    generator.display_code();

    return 0;
}
